pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/errorprone/xml/BrokenNullCheck.xml

```diff
@@ -150,40 +163,114 @@ public class Foo {
         <description>Assignments in checks are OK</description>
         <expected-problems>0</expected-problems>
         <code><![CDATA[
-public class Foo {
-    void bar() {
+public class Book {
+    Book[] books;
+    Book[] getBook(int i) {return books;}
+    void bar(int janeEyre) {
         if (books == null &&
             (books = getBook(janeEyre)) == null) { }
     }
 }
         ]]></code>
     </test-code>
 
-    <!-- A rare edge case that is technically a false positive,
-    although the code is senseless. Will not fix. -->
-<!--<test-code>
+    <test-code>
         <description>Useless code, but it will not throw a NullPointerException</description>
         <expected-problems>0</expected-problems>
         <code><![CDATA[
-public class Foo {
-    void bar() {
+public class Family {
+    String plain, bold;
+    void bar(Family family) {
         if (family.plain == null && family.bold == null &&
            family.plain == null && family.bold == null) { }
     }
 }
         ]]></code>
-    </test-code> -->
+    </test-code>
 
     <test-code>
         <description>Similar method calls with different arguments are OK</description>
         <expected-problems>0</expected-problems>
         <code><![CDATA[
-public class Foo {
-    void bar() {
+import my.Jane;
+import my.Eyre;
+public class Library {
+    String[] getBook(Class<?> i) {return null;}
+    void bar(Library library) {
         if (library.getBooks(Jane.class) != null ||
             library.getBooks(Eyre.class) != null) { }
     }
 }
         ]]></code>
     </test-code>
+    <test-code>
+        <description>#3071 BrokenNullCheck FP</description>
+        <expected-problems>0</expected-problems>
+        <code><![CDATA[
+            class Test{
+                Object obj = new Object(), objectWithDifferentName = obj;
+
+                public void testPMD(){
+                    Object obj = new Object();
+                    String str = "str";
+                    Boolean boo = null;
+
+                    //Case 1. : An Alarm is generated
+                    if(obj == null && !(obj instanceof String));
+
+                    //Case 2. : Alarms are generated for each line.
+                    if(boo == null && boo != true);
```

Comment:
We have to report this - boo is null and for comparison with "true", auto-unboxing takes place, which throws a null pointer exception. This is exactly the case, the rule is supposed to detect...

So, this case 2 is not a false positive....